wikiget.1
.\" Automatically generated by Pandoc 3.1.8
.\"
.TH "WIKIGET" "1" "October 5, 2023" "Version 0.5.1" "Wikiget User Manual"
.SH NAME
wikiget - download files from MediaWiki sites
.SH SYNOPSIS
.PP
\f[B]wikiget\f[R] [\f[I]options\f[R]] \f[I]FILE\f[R]
.PD 0
.P
.PD
\f[B]wikiget\f[R] [\f[I]options\f[R]] [-\f[B]a\f[R]|--\f[B]batch\f[R]]
\f[I]BATCHFILE\f[R]
.PD 0
.P
.PD
\f[B]wikiget\f[R] [-\f[B]V\f[R]|--\f[B]version\f[R]]
.PD 0
.P
.PD
\f[B]wikiget\f[R] [-\f[B]h\f[R]|--\f[B]help\f[R]]
.SH DESCRIPTION
Something like \f[B]wget\f[R](1) for downloading a file from MediaWiki
sites (like Wikipedia or Wikimedia Commons) using only the file name or
the URL of its description page.
.SH OPTIONS
.TP
\f[I]FILE\f[R]
The file to be downloaded.
If \f[I]FILE\f[R] is in the form \f[I]File:Example.jpg\f[R] or
\f[I]Image:Example.jpg\f[R], it will be fetched from the default site,
which is \[lq]commons.wikimedia.org\[rq].
If it\[cq]s the fully-qualified URL of a file description page, like
\f[I]https://en.wikipedia.org/wiki/File:Example.jpg\f[R], the file is
fetched from the site in the URL, in this case
\[lq]en.wikipedia.org\[rq].
.TP
\f[I]BATCHFILE\f[R]
In batch download mode (activated with -\f[B]a\f[R] or
--\f[B]batch\f[R]), this is a text file containing multiple file names
or URLs to be downloaded, one per line.
If an error is encountered during download, execution stops immediately
and the offending filename is printed.
.TP
-\f[B]s\f[R], --\f[B]site\f[R] \f[I]SITE\f[R]
MediaWiki site to download from.
Will not have any effect if the full URL is given in the \f[I]FILE\f[R]
parameter.
.TP
-\f[B]P\f[R], --\f[B]path\f[R] \f[I]PATH\f[R]
Script path for the wiki, where \[lq]index.php\[rq] and
\[lq]api.php\[rq] live.
On Wikimedia sites, it\[cq]s \[lq]/w/\[rq], the default, but other sites
may use \[lq]/\[rq] or something else entirely.
.TP
-\f[B]u\f[R], --\f[B]username\f[R] \f[I]USERNAME\f[R]
Username for private wikis that require a login even for read access.
.TP
-\f[B]p\f[R], --\f[B]password\f[R] \f[I]PASSWORD\f[R]
Password for private wikis that require a login even for read access.
.TP
-\f[B]o\f[R], --\f[B]output\f[R] \f[I]OUTPUT\f[R]
By default, the output filename is the same as the remote filename
(without the File: or Image: prefix), but this can be changed with this
option.
.TP
-\f[B]l\f[R], --\f[B]logfile\f[R] \f[I]LOGFILE\f[R]
Specify a logfile, which will contain detailed information about the
download process.
If the logfile already exists, new log information is appended to it.
.TP
-\f[B]f\f[R], --\f[B]force\f[R]
Force existing files to be overwritten.
.TP
-\f[B]a\f[R], --\f[B]batch\f[R]
If this flag is set, \f[B]wikiget\f[R] will run in batch download mode
(see \f[I]BATCHFILE\f[R]).
.TP
-\f[B]j\f[R], --\f[B]threads\f[R]
Number of parallel downloads to attempt in batch mode.
This option has no effect if -\f[B]a\f[R] is not also set.
.TP
-\f[B]v\f[R], --\f[B]verbose\f[R]
Print additional information, such as the site used and the full URL of
the file.
Additional invocations will increase the level of detail.
.TP
-\f[B]q\f[R], --\f[B]quiet\f[R]
Silence warnings and minimize printed output.
.TP
-\f[B]V\f[R], --\f[B]version\f[R]
Print the version number of the program.
.TP
-\f[B]h\f[R], --\f[B]help\f[R]
Print a brief summary of these options.
.SH EXAMPLES
.IP
.EX
wikiget File:Example.jpg
wikiget --site en.wikipedia.org File:Example.jpg
wikiget https://en.wikipedia.org/wiki/File:Example.jpg -o test.jpg
.EE
.SH BUG REPORTS
https://github.com/clpo13/wikiget/issues
.SH LICENSE
Copyright (C) 2018-2023 Cody Logan and contributors
.PP
This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.
.PP
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along
with this program.
If not, see https://www.gnu.org/licenses/.
.SH AUTHORS
Cody Logan <clpo13@gmail.com>.


src/wikiget/wikiget.py
# wikiget - CLI tool for downloading files from Wikimedia sites
# Copyright (C) 2018-2023 Cody Logan and contributors
# SPDX-License-Identifier: GPL-3.0-or-later
#
# Wikiget is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Wikiget is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Wikiget. If not, see <https://www.gnu.org/licenses/>.

import argparse
import logging
import sys
from concurrent.futures import ThreadPoolExecutor

import wikiget
from wikiget.dl import download, prep_download


def construct_parser():
    parser = argparse.ArgumentParser(
        description="""
        A tool for downloading files from MediaWiki sites using the file name or
        description page URL
        """,
        epilog="""
        Copyright (C) 2018-2023 Cody Logan and contributors. License GPLv3+: GNU GPL
        version 3 or later <http://www.gnu.org/licenses/gpl.html>. This is free
        software; you are free to change and redistribute it under certain conditions.
        There is NO WARRANTY, to the extent permitted by law.
        """,
    )
    parser.add_argument(
        "FILE",
        help="""
        name of the file to download with the File: prefix, or the URL of its file
        description page
        """,
    )
    parser.add_argument(
        "-V",
        "--version",
        action="version",
        version=f"%(prog)s {wikiget.wikiget_version}",
    )
    message_options = parser.add_mutually_exclusive_group()
    message_options.add_argument(
        "-q", "--quiet", help="suppress warning messages", action="store_true"
    )
    message_options.add_argument(
        "-v",
        "--verbose",
        help="print detailed information; use -vv for even more detail",
        action="count",
        default=0,
    )
    parser.add_argument(
        "-f", "--force", help="force overwriting existing files", action="store_true"
    )
    parser.add_argument(
        "-s",
        "--site",
        default=wikiget.DEFAULT_SITE,
        help="MediaWiki site to download from (default: %(default)s)",
    )
    parser.add_argument(
        "-P",
        "--path",
        default=wikiget.DEFAULT_PATH,
        help="MediaWiki site path, where api.php is located (default: %(default)s)",
    )
    parser.add_argument(
        "-u",
        "--username",
        default="",
        help="MediaWiki site username, for private wikis",
    )
    parser.add_argument(
        "-p",
        "--password",
        default="",
        help="MediaWiki site password, for private wikis",
    )
    output_options = parser.add_mutually_exclusive_group()
    output_options.add_argument("-o", "--output", help="write download to OUTPUT")
    output_options.add_argument(
        "-a",
        "--batch",
        help="treat FILE as a textfile containing multiple files to download, one URL "
        "or filename per line",
        action="store_true",
    )
    parser.add_argument(
        "-l", "--logfile", default="", help="save log output to LOGFILE"
    )
    parser.add_argument(
        "-j",
        "--threads",
        default=1,
        help="number of parallel downloads to attempt in batch mode",
        type=int,
    )

    return parser


def main():
    parser = construct_parser()
    args = parser.parse_args()

    loglevel = logging.WARNING
    if args.verbose >= wikiget.VERY_VERBOSE:
        # this includes API and library messages
        loglevel = logging.DEBUG
    elif args.verbose >= wikiget.STD_VERBOSE:
        loglevel = logging.INFO
    elif args.quiet:
        loglevel = logging.ERROR

    # configure logging:
    # console log level is set via -v, -vv, and -q options;
    # file log level is always info (TODO: add debug option)
    base_format = "%(threadName)s - %(message)s"
    log_format = "[%(levelname)s] " + base_format
    if args.logfile:
        # log to console and file
        logging.basicConfig(
            level=logging.INFO,
            format="%(asctime)s [%(levelname)-7s] " + base_format,
            filename=args.logfile,
        )

        console = logging.StreamHandler()
        # TODO: even when loglevel is set to logging.DEBUG, debug messages aren't
        # printing to console
        console.setLevel(loglevel)
        console.setFormatter(logging.Formatter(log_format))
        logging.getLogger("").addHandler(console)
    else:
        # log only to console
        logging.basicConfig(level=loglevel, format=log_format)

    # log events are appended to the file if it already exists, so note the start of a
    # new download session
    logging.info(f"Starting download session using wikiget {wikiget.wikiget_version}")
    # logging.info(f"Log level is set to {loglevel}")

    if args.batch:
        # batch download mode
        input_file = args.FILE
        dl_list = []

        logging.info(f"Using batch file '{input_file}'.")

        try:
            fd = open(input_file)
        except OSError as e:
            logging.error(
                "File could not be read. The following error was encountered:"
            )
            logging.error(e)
            sys.exit(1)
        else:
            with fd:
                # store file contents in memory in case something happens to the file
                # while we're downloading
                for _, line in enumerate(fd):
                    dl_list.append(line)

        # TODO: validate file contents before download process starts
        with ThreadPoolExecutor(
            max_workers=args.threads,
            thread_name_prefix="download",
        ) as executor:
            futures = []
            for line_num, line in enumerate(dl_list, start=1):
                url = line.strip()
                # keep track of batch file line numbers for debugging/logging purposes
                logging.info(f"Downloading '{url}' at line {line_num}:")
                file = prep_download(url, args)
                future = executor.submit(download, file, args)
                futures.append(future)
            # wait for downloads to finish
            for future in futures:
                future.result()
    else:
        # single download mode
        file = prep_download(args.FILE, args)
        download(file, args)
